employee/employee-cinvoke.c

```diff
@@ -0,0 +1,95 @@
+/* This program  atomically unseal a pair of capabilities
+ * (code and data) by calling `ldpblr`, i.e.
+ * "Load Pair of capabilities and Branch with Link".
+ * See "ARM Architecture Reference Manual Supplement Morello"
+ * (https://developer.arm.com/documentation/ddi0606/latest).
+ */
+
+#include "../include/common.h"
+#include <assert.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/sysctl.h>
+
+struct cheri_object
+{
+	void *__capability datacap;
+	void *__capability codecap;
+};
+
+void print_salary(uint8_t salary);
+void invoke(struct cheri_object * pair);
+void *__capability seal_immediate(void *__capability cap);
+
+
+int main()
+{
+
+	// // Request special capability `sealcap` from the operating system
+	// // in order to use it as key to seal our capabilities
+	// void *__capability sealcap;
+	// size_t sealcap_size = sizeof(sealcap);
+	// if (sysctlbyname("security.cheri.sealcap", &sealcap, &sealcap_size, NULL, 0) < 0)
+	// {
+	// 	error("Fatal error. Cannot get `securiy.cheri.sealcap`.");
+	// 	exit(1);
+	// }
+	// assert(cheri_perms_get(sealcap) & CHERI_PERM_SEAL);
+	struct cheri_object * obj = (struct cheri_object *) malloc(sizeof(struct cheri_object)); 
+	uint8_t *small_salary = (uint8_t *) malloc(sizeof(uint8_t));
+	pp_cap(&print_salary);
+	pp_cap(small_salary);
+
+	// Seal the entire struct
+	// obj->codecap = &print_salary;
+	// obj->datacap = small_salary;
+	// // Seal `cheri_object` using previously requested `sealcap`
+	// obj = (struct cheri_object *) cheri_seal(obj, sealcap);
+	// assert(cheri_is_sealed(obj->codecap));
+	// assert(cheri_is_sealed(obj));
+
+	
+	obj->datacap = seal_immediate(small_salary);
+	obj->codecap = seal_immediate(&print_salary);
+	printf("After sealing...\n");
+	pp_cap(&print_salary);
+	pp_cap(small_salary);
+	assert(cheri_is_sealed(obj->datacap));
+	assert(cheri_is_sealed(obj->codecap));
+
+	invoke(obj);
+
+	return 0;
+}
+
+inline void invoke(struct cheri_object * cheri_obj){
+	// Data for the function
+	// uint8_t * datacap = (uint8_t *) cheri_obj->datacap;
+	// Function pointer (print_salary) of target function
+	// void (*codecap)(void) =  cheri_obj->codecap;
+	asm(
+		"ldpblr c0, [%w[pair]]\n\t"
+		"ret\n\t"
```

Comment:
We shouldn't return here. (Sorry if this is my poor suggestion from an earlier comment.)

This will actually cause an infinite loop, since the `ldpblr` sets `lr` to this address, then this `ret` jumps to itself.

Just delete this instruction, add `clr` to the clobber list (as shown below), and it'll be fine.

Reply:
After adding the registers to the clobbered list, `qemu` kicks me out:

![kicked](https://user-images.githubusercontent.com/58388250/132682410-c0e562bd-cc61-4724-b704-985755d7adc1.png)

Updates In 082b1c0.

Reply:
Is this related to the fact you mentioned, i.e. the `datacap` is not unsealed/cannot be? Do we need to set the OType somehow or am again wrong and tied to the abstract specs?

Reply:
First, I suggest fixing the problems in the sealing function. I think that will resolve this. With the assembly explicitly writing to `c1`, I think the sealed result will just be dropped (because the compiler doesn't know about it).

In the meantime, I'll try getting `ldpblr` to work with an unsealed capability. If I read the specification correctly, `ldpblr` should accept unsealed inputs, so either I've misunderstood that, or that assertion is wrong.

Reply:
> Is this related to the issue you mentioned, i.e. the `datacap` is not unsealed/cannot be? Do we need to set the OType somehow or am again wrong and tied to the abstract specs?

Sorry, I missed this somehow. Yes, I think it's related, but it's hard to follow the threads now.
"Sealed" and "has a non-zero OType" are the same thing, effectively. You don't need to set it a second time.

Reply:
> > Is this related to the issue you mentioned, i.e. the `datacap` is not unsealed/cannot be? Do we need to set the OType somehow or am again wrong and tied to the abstract specs?
> 
> Sorry, I missed this somehow. Yes, I think it's related, but it's hard to follow the threads now.

Oh, no, I am sorry. I had to do small modifications comment by comment and link them to specific comments. My mistake.

Reply:
> In the meantime, I'll try getting `ldpblr` to work with an unsealed capability. 

Done in 2c424aa. It works with unsealed capabilities.

Reply:
Saying "it works", it's a mistake, I think. It always print a dirty value of "16". Is using a function with a data argument allowed b the specs? I think `datacap` is supposed to be accessed by the function as part of a "class" (in OO terms) not passed as an argument.